mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -218,7 +218,7 @@ private object IDFModel {
           newValues(k) = values(k) * idf(indices(k))
           k += 1
         }
-        Vectors.sparse(n, indices, newValues)
+        Vectors.sparse(n, indices, newValues).toSparse
```

Comment:
Makes some sense but `Vectors.sparse` is already doing some expensive operations like a sort. A filter doesn't seem like much extra. 

I get that someone setting 0 directly into a sparse vector might want it to literally record it, but is that someone's expectation in passing it into a factory method?

Anyway maybe you're on to something here, just sounds funny that we have to separately count non-zero elements